src/HealthGPS/kevin_hall_model.cpp

```diff
@@ -81,74 +94,114 @@ void KevinHallModel::update_risk_factors(RuntimeContext &context) {
+        auto *messagePrt = dynamic_cast<KevinHallAdjustmentMessage *>(basePtr.get());
+        if (!messagePrt) {
+            throw core::HgpsException(
+                "Simulation out of sync, failed to receive a Kevin Hall adjustments message");
+        }
 
-    // Final run.
+        adjustments = messagePrt->data();
+    }
+
+    // Adjust: different model for children and adults.
     for (auto &person : context.population()) {
         // Ignore if inactive.
         if (!person.is_active()) {
             continue;
         }
 
-        // TODO: Simulate person and update risk factors.
-        simulate_person(person, shift);
-        // SimulatePersonState state = simulate_person(person, shift);
-        // person.risk_factors["Height"_id] = state.H;
-        // person.risk_factors["Weight"_id] = state.BW;
-        // person.risk_factors["PhysicalActivity"_id] = state.PAL;
-        // person.risk_factors["RestingMetabolicRate"_id] = state.RMR;
-        // person.risk_factors["BodyFat"_id] = state.F;
-        // person.risk_factors["LeanTissue"_id] = state.L;
-        // person.risk_factors["ExtracellularFluid"_id] = state.ECF;
-        // person.risk_factors["Glycogen"_id] = state.G;
-        // person.risk_factors["EnergyExpenditure"_id] = state.EE;
-        // person.risk_factors["EnergyIntake"_id] = state.EI;
+        auto &adjustment = adjustments.at(person.gender, person.age);
+
+        if (person.age < kevin_hall_age_min) {
+            initialise_kevin_hall_state(person, adjustment);
+        } else {
+            kevin_hall_adjust(person, adjustment);
+        }
+    }
+
+    // Baseline scenario: send adjustments to intervention scenario.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        context.scenario().channel().send(std::make_unique<KevinHallAdjustmentMessage>(
+            context.current_run(), context.time_now(), std::move(adjustments)));
     }
 }
 
+KevinHallAdjustmentTable
+KevinHallModel::compute_kevin_hall_adjustments(Population &population) const {
+    auto expected = get_risk_factor_expected();
+    auto weight_means = compute_mean_weight(population);
+
+    // Compute adjustments.
+    auto adjustments = KevinHallAdjustmentTable{};
+    for (const auto &[sex, weight_means_by_sex] : weight_means) {
+        adjustments.emplace_row(sex, std::unordered_map<int, double>{});
+        for (const auto &[age, weight_mean] : weight_means_by_sex) {
+            double weight_expected = expected.at(sex, "Weight"_id).at(age);
+            adjustments.at(sex)[age] = weight_expected - weight_mean;
+        }
+    }
+
+    return adjustments;
+}
+
 void KevinHallModel::initialise_nutrient_intakes(Person &person) const {
 
     // Initialise nutrient intakes.
     set_nutrient_intakes(person);
 
-    // TODO: set old value to new value (only some nutrients)
+    // Start with previous = current.
+    double carbohydrate = person.risk_factors.at("Carbohydrate"_id);
```

Comment:
This is just a style question: why sometimes do you use `auto` and in others the explicit type?

Reply:
Good question.. I think I try to use the type when I want to make it obvious what it is, e.g. double, but use auto where it's a bit long type, like `map<int, map<Identifier, double>>`, or where it's obvious from the context (like `static_cast<int>(...)` or `get_complex()` etcetera. I think clang-tidy even enforces auto use with casting. I'm not sure if there are any useful standard guidelines though.

Reply:
You can even make `clang-tidy` enforce use of `auto` *everywhere* if you're particularly keen on it :laughing:. I agree with James though that it's usually nice to be explicit for short type names though.